### DISCUSSION TITLE
Running into a problem with templates, not sure of reason

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#147] [Q&A] [answered]
I just made a local copy of the C# grammar fault-tolerant with
```
FAULT_TOLERANT;
FAULT_TOLERANT_DEFAULT=false;
```
and now the Python templates give an indentation error. The problem is in this area:
https://github.com/congo-cc/congo-parser-generator/blob/8f2036b01d6ee48b71eaf20ec852aee09500146e/src/templates/python/parser_productions.inc.ftl#L140-L148
and the relevant part of the invoked macro is
https://github.com/congo-cc/congo-parser-generator/blob/8f2036b01d6ee48b71eaf20ec852aee09500146e/src/templates/python/common_utils.inc.ftl#L85-L96
I generally use this approach everywhere and have been for a very long time. I'm not sure why it should now start failing, so I'm wondering if I'm doing something wrong with my use of FreeMarker. The generated code is:
```python
            previousLexicalState4 = self.token_source.lexical_state
            self.token_source.reset(self.last_consumed_token, LexicalState.IN_MULTI_INTERPOLATION)
            try:
            if self.pending_recovery:
                #   if self.debug_fault_tolerant:
                #       logger.info('Re-synching to expansion at: CSharp.ccc:2412:4')
                self.recoverΣMultiNonInterpolatedText()
                # Code for Terminal specified at CSharp.ccc:2412:4
                follow_set12 = None
                if self.outer_follow_set is not None:
                    follow_set12 = set(self.follow_setΣCSharp_cccΣ2412Σ4) | self.outer_follow_set
                self.consume_token(NON_INTERPOLATED_TEXT2, False, follow_set12)
            finally:
```
It certainly looks as if the nested content (the first `if` statement) isn't being indented correctly, but I can't see why. Perhaps you can help, @revusky?

**Comment by revusky** [comment]
I don't honestly know. The truth is that you surely undertand this python indentation stuff (that you did) better than I do! To tell the truth, I think we need a better, more general solution to this problem. Some months back (or maybe a whole year at this point) I thought I saw a good way of dealing with this, but I got totally distracted. With Java, we output all the code willy-nilly basically and then it goes through a pretty-printing pass at the very end. Of course, the thing is that, with Java, even if you get some of the indentation wrong, the code still compiles okay. That is not the case with Python.

**Comment by vsajip** [comment]
Well, it's not really a Python issue, it's a template issue. I just want the nested part to be indented, and `[#nested indent + 4]` is used in several places in the template and produces the expected results. It's just a question of wanting the nested content to be indented, and it generally is, but not in this case. It could be anything other than Python, but the question is how to get a certain result in the output from a FreeMarker template. I think it's either an error on my part in how I'm using FreeMarker (though the same  `[#nested indent + 4]` pattern works in other places), or it's some FreeMarker issue which just surfaces in this one case for some reason.

**Comment by vsajip** [comment]
Specifically, the output should be
```python
            try:
                if self.pending_recovery:
                    #   if self.debug_fault_tolerant:
                    #       logger.info('Re-synching to expansion at: CSharp.ccc:2412:4')
                    self.recoverΣMultiNonInterpolatedText()
                # Code for Terminal specified at CSharp.ccc:2412:4
                follow_set12 = None
                if self.outer_follow_set is not None:
                    follow_set12 = set(self.follow_setΣCSharp_cccΣ2412Σ4) | self.outer_follow_set
                self.consume_token(NON_INTERPOLATED_TEXT2, False, follow_set12)
            finally:
```
Where the nested content is the first four lines in the `try` statement.

**Comment by adMartem** [comment, marked as answer]
I think what is happening here is that the `is` variable is being evaluated in the context of the outer macro invoking the `#nested `directive.  I experimented and found the `indent` is correct, but `is` is not ,in the context of the nested lines.  When I substitute the `""?right_pad(indent) `expression for `is`, it works.

**Comment by adMartem** [comment]
I ran across [this](https://github.com/freemarker/freemarker3/blob/e15ebbe359ad1584fd615ae9265ca583bf8ab8cb/src/parser/Directives.inc.ccc#L551), and I think it is pertinent.

**Comment by adMartem** [comment]
It bothered me that I could swear I did what you did in the ParserProductions template for python (the `[#nested indent+4]` (no spaces) in `TreeBuildingAndRecovery` macro).  But the difference is that I didn't generate any code directly in the `TreeBuildingAndRecovery` macro, but rather I just passed the biased `indent` value to macros inside the nested block that actually resolved `is` from their `indent` actual arguments.

**Comment by adMartem** [comment]
While I was in there making other changes I went ahead and patched the 4 lines as described above so that they now work, but it could probably be made prettier.  This is in PR #148.
